File: Python/Programmers/Level 2/숫자의 표현.py
# 연속되는 자연수의 합이 n이 되려면 연속되는 자연수 중 제일 작은 수가 n//2보다 작거나 같아야함
# 따라서 모든 수를 탐색하는 것이 아닌 이분탐색 사용하고 마지막에 +1

# 이중 반복문 사용하여 비교
# def solution(n):
#     answer, idx = 0, 1
#     while idx <= n:
#         tmp = 0
#         for i in range(idx, n+1): # idx부터 n까지 연속된 자연수의 합이 n이 되는지 체크
#             tmp += i
#             if tmp > n: # tmp가 n과 같아지지 않고 크다면 break
#                 break
#             elif tmp == n: # tmp가 n과 같아진다면 answer 값 1 증가시키고 break
#                 answer += 1
#                 break
#         idx += 1 # idx값 +1 증가시켜서 연속된 자연수를 모두 파악

#     return answer

def solution(n):
    # 연속되는 자연수 중 제일 작은 수 first, 연속되는 자연수의 합이 되는지 확인하기 위한 check, 연속되는 자연수를 표현할 idx 선언
    first, check, idx = 1, 0, 0
    answer = 0
    while first < (n//2+1):
        check += (first + idx) # check에 first와 idx 값을 더해나감
        idx += 1 # 숫자가 연속해서 커져야하므로 idx += 1
        
        if check >= n: # check가 n보다 크거나 같으몀
            if check == n: # 그 중 check가 n과 같다면 answer += 1
                answer += 1
            first += 1 # first 값 +1
            check, idx = 0, 0 # check, idx 값 초기화
    
    return answer + 1 # 이분 탐색이므로 마지막 n은 체크를 안함 따라서 +1

File: JavaScript/Programmers/Level 2/숫자의 표현.js
// 1부터 연속된 수의 합이 n이 되는지 확인하면 된다.
// 연속된 수가 되려면 n / 2 값보다 작아야함

// function solution(n) {
//     let answer = 0;
//     for (let i = 1; i < n / 2; i++) { // 1부터 연속된 수의 합이 n이 되는지 확인, 초기 시작 숫자
//         let tmp = 0;
//         for (let j = i; j < n + 1; j++) { // 더해지는 숫자
//             tmp += j; // j의 값을 더해나감
//             if (tmp >= n) { // tmp 값이 n보다 크거나 같으면
//                 (tmp === n) && answer++; // 그 중 tmp 값이 n과 같다면 answer 카운팅
//                 break;
//             }
//         }
//     }
    
//     return answer + 1; // n / 2 연산으로 n을 고려하지 않았기 때문에 answer + 1 값 return
// }

// 중첩 for문 대신 while문 사용, 시간은 비슷함

// function solution(n) {
//     let answer = 0, init = 1, currentNum = 0, sum = 0;
//     while (init < n / 2) { // 초기 시작 값인 init이 n / 2보다 작을 동안 반복
//         sum += (init + currentNum); // sum에 init부터 연속된 숫자의 합을 구함
//         if (sum >= n) { // sum이 n 이상이면
//             (sum === n) && answer++; // 그 중 sum이 n과 같다면 answer 카운팅
//             init++; // init 값 1증가
//             // sum과 currentNum 값 초기화
//             sum = 0;
//             currentNum = 0;
//             continue;
//         }
//         currentNum++; // currentNum 값 1증가
//     }    
    
//     return answer + 1;
// }

// 규칙을 찾아보면 n의 약수 중 홀수의 개수가 정답이 됨
// n의 약수는 n의 제곱근 값을 기준으로 대칭됨

function solution(n) {
    let answer = 0;
    for (let i = 1; i <= (n ** 0.5); i++) { // 1부터 n까지 n번 연산을 수행하지 않고 n의 제곱근까지만 연산 수행
        if (n % i === 0) { // i가 약수라면
            const quotient = +(n / i); // i로 나누었을 때 몫도 약수
            if (quotient % 2 === 1) { // 몫이 홀수라면 카운팅
                answer++;
            }
            // 예를 들어, n이 16이라면 약수는 1, 2, 4, 8, 16이 됨. 이때 4는 제곱근 값이기 때문에 중복 카운팅을 막기 위해 (quotient !== i) 조건 추가
            if ((quotient !== i) && (i % 2 === 1)) { // i도 홀수라면 
                answer++;
            }
        }
    }
    
    return answer;
}
